docs/design/generic_data_ingestion.md

```diff
@@ -0,0 +1,325 @@
+    `__run_graph__` tag name, these should be name-scoped by kind: a
+    user-provided tag `my_keras_model` should become
+    `__model_graph__/my_keras_model`. (That is, the tag name encodes a
+    discriminated union.)
+-   **hparams:** Could be ingested as a blob sequence or tensor (all the data is
+    actually in the tag name and summary metadata; the tensor content is empty).
+    But we really want hparams to be some special run-level metadata rather than
+    summaries, so maybe we hold off on doing anything for now.
+
+\* We could also opt to drop the dimensions here, as the frontend doesn’t
+actually need them, but (a) [we may want to use them later][image-dims], and (b)
+this would basically be an orthogonal change. If the Storzy blobstore is
+content-addressed, the extra storage is effectively free, as each “dimension”
+file will only be stored once across all users. If it’s a unique-owner store,
+we’re creating 67% more files per standard 3-sample summary, though the actual
+file content bytes are of course negligible. Bandwidth costs at runtime should
+be zero in any case, because the Python plugin just won’t call read_blob on
+these two elements.
+
+[audio-labels]: https://github.com/tensorflow/tensorboard/blob/a0c1def1e3a3725909f9fd59ee5d50fda81bb7ea/tensorboard/plugins/audio/summary_v2.py#L21
+[audio-labels-request]: https://github.com/tensorflow/tensorboard/issues/296#issuecomment-320146370
+[histo-shape]: https://github.com/tensorflow/tensorboard/blob/a0c1def1e3a3725909f9fd59ee5d50fda81bb7ea/tensorboard/plugins/histogram/summary.py#L17-L25
+[image-dims]: https://github.com/tensorflow/tensorboard/pull/3009#pullrequestreview-329605910
+[pr-curves-shape]: https://github.com/tensorflow/tensorboard/blob/4ee9ae8a31524131eb56f7a1dc4aa09d7d186f10/tensorboard/plugins/pr_curve/summary.py#L523-L532
+
+## Mechanism structure
+
+We propose a declarative mechanism by which each summary on disk declares its
+appropriate data class.
+
+The `SummaryMetadata` proto will gain a new field `DataClass` of enum type,
+which may be “unknown” (default), “scalar”, “tensor”, or “blob sequence”. All
+newly written data must set this field to be compatible with data providers. A
+read-time compat layer will transform known first-party plugin data into this
+new form, analogous to the existing `data_compat.py` layer that transparently
+migrates v1-style summaries to v2-style summaries.
+
+### Protocol buffer changes
+
+In `summary.proto`, add a new enumerated type:
+
+```proto
+enum DataClass {
+    DATA_CLASS_UNKNOWN = 0;
+    DATA_CLASS_SCALAR = 1;
+    DATA_CLASS_TENSOR = 2;
+    DATA_CLASS_BLOB_SEQUENCE = 3;
+}
+```
+
+Then, extend the existing `SummaryMetadata` to include a field of this type:
+
+```proto
+message SummaryMetadata {  // (extend)
+    DataClass data_class = 4;
+}
+```
+
+The value of `data_class` on a tag’s summary metadata, whether explicitly set or
+implicitly zero, establishes constraints on the tensor values for all Values of
```

Comment:
nit strange capital V in Values

Reply:
Thanks! This is correct, actually—it’s referring to the `Value` protos
in the `Summary.value` repeated field. But it should have code
formatting to make that clear: done.